test/src/containerTestBase.ts

```diff
@@ -0,0 +1,139 @@
+import { resolve, getDevContainerConfigPathIn, readDevContainerConfigFile } from 'remote-containers/src/node/configContainer';
+import { ResolverResult } from 'remote-containers/src/node/utils';
+
+
+// Dev Container schema
+log('debug', 'Downloading devcontainer.json schema');
+const schemaPromise = getStringFromUri(getConfig('devContainerSchemaUri'))
+	.then((schemaString: string) => { 
+		return jsonc.parse(schemaString); 
+	});
+
+// Main test description
+export function describeTest(description: string, rootFolder: string, definitionList: string[]) {
+
+	definitionList.forEach((definition: string) => {
+		describe(`${description}: ${definition}`, function () {
+			const definitionPath = path.join(rootFolder, definition);
+			let devContainer: ResolverResult;
+	
+			// Validate devcontainer.json
+			it('devcontainer.json should be valid', async function () {
+				this.slow(5000);
+				assert.isTrue(validateDevContainerJson(definitionPath, await schemaPromise));
+			});
+	
+			// Some definitions won't build since they are templates, skip them
+			if (getConfig('skipDefinitionBuild').indexOf(definition) > -1) {
+				return true;
+			}
+	
+			// Build or pull image specified by devcontainer.json
+			it('should start', async function () {
+				this.slow(120000);
+				this.timeout(0);
+	
+				devContainer = await testContainerResolver(definitionPath);
+				assert.equal(await fetchAgentCommit(devContainer.resolvedAuthority), devContainer.params.product.commit);
+	
+				return true;
+			});
+	
+			it('should execute', async function () {
+				this.slow(120000);
+				this.timeout(0);
+	
+				const result = await execTestScript(devContainer, path.resolve(rootFolder, '..'));
+				assert.isNull(result.error, result.error + ': ' + result.stderr);
+	
+				return true;
+			});
+	
+			it('should clean up', async function () {
+				this.slow(120000);
+				this.timeout(0);
+	
+				const result = await cleanUpTest(devContainer);
+				assert.isNull(result.error, result.error + ': ' + result.stderr);
+	
+				return true;
+			});
+	
+			return true;
+		});
+	});
+}
+
+
+export function validateDevContainerJson(definitionPath: string, schema: any) {
+	log('debug', `Validating devcontainer.json at ${definitionPath}`);
+	const jsonPath = path.join(definitionPath, '.devcontainer', 'devcontainer.json');
+	const json = fs.existsSync(jsonPath) ? jsonc.readSync(jsonPath) :
+		jsonc.readSync(path.join(definitionPath, '.devcontainer.json'));
+
+	return tv4.validate(json, schema);
+}
+
+export async function testContainerResolver(definitionPath: string): Promise<ResolverResult> {
+
+	log('debug', `Finding and reading devcontainer.json at ${definitionPath}`);
+	const configPath = await getDevContainerConfigPathIn(definitionPath);
+	const config: any = configPath && await readDevContainerConfigFile(configPath);
+
+	const params = await createTestParams();
+	params.getExtensionsToInstall = () => config.extensions;
```

Comment:
If `configPath` is undefined, line 93 will still set config to `undefined`. Should the error being thrown have more meaning than the possible `cannot read property "extensions" of undefined`?